sockeye/lexical_constraints.py

```diff
@@ -0,0 +1,399 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You may not
+# use this file except in compliance with the License. A copy of the License
+# is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is distributed on
+# an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+import logging
+import time
+import copy
+import re
+from typing import Dict, List, NamedTuple, Tuple
+from operator import attrgetter
+
+from . import constants as C
+from . import utils
+
+import mxnet as mx
+import numpy as np
+
+logger = logging.getLogger(__name__)
+
+RawConstraintList = List[List[int]]
+
+class ConstrainedHypothesis:
+    """
+    Represents a set of words and phrases that must appear in the output.
+    A constraint is of two types: sequence or non-sequence.
+    A non-sequence constraint is a single word and can therefore be followed by anything, whereas a sequence constraint must be followed by a particular word (the next word in the sequence).
+    This class also records which constraints have been met.
+
+    :param constraint_list: A list of zero or raw constraints (each represented as a list of integers).
+    :param eos_id: The end-of-sentence ID.
+    """
+    def __init__(self, constraint_list: RawConstraintList, eos_id: int) -> None:
+
+        # `constraints` records the words of the constraints, as a list (duplicates allowed).
+        # `is_sequence` is a parallel array that records, for each corresponding constraint,
+        self.constraints = []
+        self.is_sequence = []
```

Comment:
Felix's comment from above: adding `  # type: List[X]` with specified X for type checking.

Reply:
done